### PR TITLE
Convert joomla_version.rb to a Joomla mixin

### PR DESCRIPTION
**Pull request by wchen-r7**
This PR converts the joomla_version.rb to a mixin. The version info check can be useful for other Joomla modules.
## Verification
- [x] Make sure Travis is green
- [x] Prepare a Joomla box
- [x] Start msfconsole
- [x] Do: `use auxiliary/scanner/http/joomla_version`
- [x] Do: `set rhosts [IP]`
- [x] Do: `set targeturi [Path to joomla]`
- [x] Do: `run`
- [x] The module should tell you what version

And then do this:
- [x] Do: `set targeturi /bad/`
- [x] Do: `run`
- [x] The module should say "It doesn't look like Joomla is up and running at /bad"
## Demo

```
msf auxiliary(joomla_version) > set targeturi /joomla/
targeturi => /joomla/
msf auxiliary(joomla_version) > run

[*] Server: Apache/2.4.7 (Ubuntu)
[*] Joomla version: 3.4.3
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(joomla_version) > 
```

```
msf auxiliary(joomla_version) > set targeturi /bad
targeturi => /bad
rumsf auxiliary(joomla_version) > run

[-] It doesn't look like Joomla is up and running at /bad
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(joomla_version) > 
```

### CODE REVIEW

File: lib/msf/core/exploit/http/joomla/version.rb [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+module Msf::Exploit::Remote::HTTP::Joomla::Version
+
+  # Returns the Joomla version.
+  #
+  # @return [String] Joomla version (if found)
+  # @return [NilClass] No Joomla version found.
+  def joomla_version
+    files = [
+      'language/en-GB/en-GB.xml',
```
Comment (FireFart) [inline lib/msf/core/exploit/http/joomla/version.rb]: we should add `administrator/manifests/files/joomla.xml` to this list. The regexes below should already be able to parse this file.

Comment (wchen-r7) [inline lib/msf/core/exploit/http/joomla/version.rb]: Yes you're right, I'll add that.

Comment (wchen-r7) [inline lib/msf/core/exploit/http/joomla/version.rb]: And done.

**Comment by FireFart** [comment]
@wchen-r7 thx! I just took a look at some sites and many sites block access to joomla.xml and remove the generator tag. But i found that some sites have the `/administrator/`directory open, which contains a generator tag and other joomla references so it may be good to add this one to the checked paths too.
Also if you take a look at the fingerprinting methods from joomscan, there is a lot of checking going on (but it's really out of date so maybe only reliable on old versions):
http://sourceforge.net/p/joomscan/code/HEAD/tree/trunk/joomscan.pl
(search for `fprint_`)

**Comment by wchen-r7** [comment]
Thanks @FireFart, I will update #joomla_and_online? in a bit. I guess I'll lower the firing order for joomla.xml a little bit, too.  Putting it second I hope works okay.

**Comment by wchen-r7** [comment]
@FireFart It looks like to port joomscan, it will require much more time. Let me know what's reasonable for you (partially to all), and if necessary I can close this PR first and allocate a couple of months to work on this (rspes & testing & stuff).

**Comment by FireFart** [comment]
@wchen-r7 I think it's not worth the effort because it looks like joomscan only detects version 1.x :( Will test your changes in a bit!

File: lib/msf/core/exploit/http/joomla/base.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+module Msf::Exploit::Remote::HTTP::Joomla::Base
+
+  # Checks if Joomla is up and running.
+  #
+  # @return [TrueClass] Joomla is up and running.
+  # @return [FalseClass] Joomla is not up.
+  def joomla_and_online?
+    # Possible paths that we might see the generator tag.
+    paths = [ '/', '/administrator' ]
```
Comment (FireFart) [inline lib/msf/core/exploit/http/joomla/base.rb]: This should be `/administrator/` otherwise strange things and redirects happen o_O

**Comment by FireFart** [comment]
Another strange thing I just encountered: I set `RHOSTS` to a domain name, and the HTTP request is
```
GET /administrator HTTP/1.1
Host: 1.1.1.1 (changed)
```

Shouldn't the host header be the `RHOSTS` value if it's no IP and the `VHOST` option is not set?

**Comment by wchen-r7** [comment]
@FireFart As the aux module initializes, the scanner mixin will ask Rex (RangeWalker) to parse your RHOSTS, and then uses the parsed version to set RHOST and pass it to the auxiliary module. Since the HttpClient mixin relies on the RHOST datastore option, I am guessing this is the expected behavior. Please feel free to double check.

**Comment by FireFart** [comment]
Oh :( I thought it behaves as on exploit modules:
https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/http/client.rb#L170

Here the vhost/Host header is set to RHOST if it's not set by the user. As we are dealing with a possible range here, I know this can be tricky.

**Comment by FireFart** [comment]
just saw @rastating has a joomla scanner online: https://github.com/rastating/joomlavs
maybe he has some more input for determing the version

**Comment by wchen-r7** [comment]
@FireFart It looks like JoomlaVS grabs the version from 2 sources:

* The metadata tag - This one already doesn't work for me. It's looking for a version in the generator tag, but mine doesn't provide a version.
* The README.txt file - Not sure how reliable it is.

**Comment by wchen-r7** [comment]
Actually, the version found in README.txt is slightly more vague than other sources. For mine, instead of 3.4.3, the README only says 3.4.

**Comment by FireFart** [comment]
@wchen-r7 ok thx for looking into this!

**Comment by rastating** [comment]
I didn't find many good ways to fingerprint the version unfortunately, most reliable I found were the metadata and the README, but as @wchen-r7 mentioned, sometimes the README gives you a generic version rather than a specific one, which can be misleading.

I did see in an older Joomla scanner that seemed to have been shelved (I think it was possibly joomscan) that it began comparing hashes of common files that were changed in specific versions, however I avoided doing this in JoomlaVS as it could potentially lead to some very misleading results (e.g. if a file's hash ends up being the same in future versions).

I do also recall though that some versions (possibly older versions though rather than newer ones) would leave Joomla version numbers in some of its JavaScript files. All in all though, I found a lot of inconsistency to it, which makes it a bit more difficult to handle :(

Hopefully that info helps a bit :)

**Comment by wchen-r7** [comment]
@rastating I see, very good to know. Thank you!

**Comment by FireFart** [comment]
@wchen-r7 by looking at the next version, the only file containing the version is `joomla.xml`:
https://github.com/joomla/joomla-cms/compare/3.4.6...3.4.7
It looks like the language file is not updated on each release. According to https://www.gavick.com/blog/how-to-check-the-version-of-joomla the `en-GB.xml` stays on the same version between minor releases. So I think it would be best to move joomla.xml up to the top because otherwise the version of `en-GB.xml` will always be returned. If the `joomla.xml` file can not be fetched, the `en-GB.xml` file is returned.

**Comment by wchen-r7** [comment]
Ok sure. Thanks for looking into it!

**Comment by wchen-r7** [comment]
@FireFart done!

**Comment by wvu-r7** [comment]
I choose you, @FireFart!

**Comment by FireFart** [comment]
thx @wchen-r7 all working as expected! Will open a PR to include the mixin in the joomla RCE module

**Comment by wchen-r7** [comment]
Wonderful. Thank you @FireFart

**Comment by FireFart** [comment]
https://github.com/rapid7/metasploit-framework/blob/master/data/sounds/default/wonderful.wav

**Pull request merged**